win32ss/user/ntuser/class.c

```diff
@@ -330,36 +330,60 @@ void FASTCALL DestroyProcessClasses(PPROCESSINFO Process )
     }
 }
 
+_Maybe_raises_SEH_exception_
 static BOOL
 IntRegisterClassAtom(IN PUNICODE_STRING ClassName,
                      OUT RTL_ATOM *pAtom)
 {
     WCHAR szBuf[65];
-    PWSTR AtomName;
+    PWSTR AtomName = szBuf;
     NTSTATUS Status;
 
     if (ClassName->Length != 0)
     {
-        /* FIXME: Don't limit to 64 characters! Use SEH when allocating memory! */
-        if (ClassName->Length / sizeof(WCHAR) >= sizeof(szBuf) / sizeof(szBuf[0]))
+        if (ClassName->Length / sizeof(WCHAR) >= ARRAYSIZE(szBuf))
         {
-            EngSetLastError(ERROR_INVALID_PARAMETER);
-            return (RTL_ATOM)0;
+            AtomName = ExAllocatePoolWithTag(PagedPool,
+                                             ClassName->Length + sizeof(UNICODE_NULL),
+                                             TAG_STRING);
+
+            if (AtomName == NULL)
+            {
+                EngSetLastError(ERROR_OUTOFMEMORY);
+                return FALSE;
+            }
         }
 
-        RtlCopyMemory(szBuf,
-                      ClassName->Buffer,
-                      ClassName->Length);
-        szBuf[ClassName->Length / sizeof(WCHAR)] = UNICODE_NULL;
-        AtomName = szBuf;
+        _SEH2_TRY
+        {
+            RtlCopyMemory(AtomName,
+                          ClassName->Buffer,
+                          ClassName->Length);
+            AtomName[ClassName->Length / sizeof(WCHAR)] = UNICODE_NULL;
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            ExFreePoolWithTag(AtomName, TAG_STRING);
+            SetLastNtError(_SEH2_GetExceptionCode());
+            return FALSE;
+        }
+        _SEH2_END;
     }
     else
+    {
+        ASSERT(IS_ATOM(ClassName->Buffer));
         AtomName = ClassName->Buffer;
+    }
 
+    /* If this function throws, we leak AtomName */
```

Comment:
Why should it do that and why would this not be wrapped in SEH then?

Reply:
I did not want to alter too much of the functionality, so that if it would throw, we do not alter that behavior.

Reply:
So you don't want to "alter" a memory leak, just in case we rely on it? 🤔